projects/forest_firefighters/controllers/autonomous_mavic_v4/autonomous_mavic_v4.py

```diff
@@ -0,0 +1,347 @@
+                self.FireDetectionStatus = False
+                if verbose_target: print("Water dropped on fire target: {} at position {}".format(self.target_pos[0:2], self.curr_pos[0:2]))
+            else:
+                self.target_index += 1
+                if self.target_index > len(local_coord)-1: self.target_index=0
+                self.target_pos[0:2] = local_coord[self.target_index]
+                if verbose_target: print("Target reached! New target: ", self.target_pos[0:2])
+        
+        self.target_pos[2] = np.arctan2(self.target_pos[1] - self.curr_pos[1], self.target_pos[0] - self.curr_pos[0])
+        angle_left = self.target_pos[2] - self.curr_pos[5]
+        if abs(angle_left) >= 3.14157:
+            angle_left = (angle_left + 6.28318) % 6.28318
+
+        #turn the robot to the left or to the right according the value and the sign of angle_left
+        yaw_disturbance = self.MAX_YAW_DISTURBANCE*angle_left/6.28318
+        pitch_disturbance = clamp(np.log10(abs(angle_left)), self.MAX_PITCH_DISTURBANCE, 0.1) #non proportional and decruising function
+
+        if verbose_movement: 
+            distance_left = np.sqrt(((self.target_pos[0] - self.curr_pos[0]) **2) + ((self.target_pos[1] - self.curr_pos[1]) **2))
+            print("remaning angle: {:.4f}, remaning distance: {:.4f}".format(angle_left, distance_left))
+        return yaw_disturbance, pitch_disturbance
+
+    def GetRotationMatrix(self, thetaX=0, thetaY=0, thetaZ=0):
+        """Provide a rotation matrix 
+        Args:
+            thetaX: Rotation about the X axis
+            thetaY: Rotation about the Y axis
+            thetaZ: Rotation about the Z axis
+
+        Returns:
+            A 3x3 transformation matrix
+        """
+        rotX = np.array([[1, 0, 0], 
+                        [0, np.cos(thetaX), np.sin(thetaX)],
+                        [0, -np.sin(thetaX), np.cos(thetaX)]])
+        rotY = np.array([[np.cos(thetaY), 0, -np.sin(thetaY)],
+                        [0, 1, 0],
+                        [np.sin(thetaY), 0, np.cos(thetaY)]])
+        rotZ = np.array([[np.cos(thetaZ), -np.sin(thetaZ), 0], 
+                        [np.sin(thetaZ), np.cos(thetaZ), 0],
+                        [0, 0, 1]])
+        R = rotX*rotY*rotZ
+        return R
+
+    def ComputeMonoplotting(self, coord_fire, verbose=True):
+        """
+        Compute the world coordinates from the image coordinates of the fire.
+        Monoplotting based on https://stackoverflow.com/questions/68195056/finding-the-real-world-coordinates-of-an-object-from-a-camera
+        and http://sar.kangwon.ac.kr/etc/rs_note/rsnote/cp9/cp9-6.htm
+        Parameters: 
+        coord_fire (list): x,y image coordinates of the fire
+        verbose (bool): whether to print status messages or not
+        """
+        resolution = self.camera.getWidth(), self.camera.getHeight()
+        roll = self.curr_pos[3]
+        pitch = self.curr_pos[4]
+        yaw = self.curr_pos[5]
```

Comment:
This is the roll, pitch, yaw of the IMU, the ones of the camera may be different.